### PR TITLE
Tensor Train Optimiser

### PR DESCRIPTION
**Pull request by VolodyaCO**
## Description

Recently,  @ArtemStrashko gave a talk on a tensor train optimiser for continuous functions. There's a library out there for it. This PR wraps this library to expose TTOpt as an optimiser.

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [x] I have updated documentation.

### CODE REVIEW

**Comment by VolodyaCO** [comment]
`ttopt` does not have type hints, which is why mypy is complaining. I'm not sure what to do here...

File: src/orquestra/opt/optimizers/tensor_train_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+    def _preprocess_cost_function(
+        self, cost_function: CostFunction
+    ) -> _CostFunctionWithBestCandidates:
+        """
+        Wraps the cost function in a CostFunctionWithBestCandidates.
+        """
+        return _CostFunctionWithBestCandidates(
+            cost_function, self.maximum_number_of_candidates, self.bounds
+        )
```
Comment (AthenaCaesura) [inline src/orquestra/opt/optimizers/tensor_train_optimizer.py:216]: Where is this function being used? It's private, right? So it shouldn't be used somewhere in this PR?

File: src/orquestra/opt/optimizers/tensor_train_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
+from typing import Callable, List, Sequence, Tuple, Union
+
+import numpy as np
+from scipy.optimize import OptimizeResult
+from ttopt import TTOpt
+
+from orquestra.opt.api.functions import CallableWithGradient
+from orquestra.opt.optimizers.pso.continuous_pso_optimizer import (
+    Bounds,  # TODO: where should these Bounds live?
```
[the remark this answers is not in the document]
Comment (VolodyaCO) [reply src/orquestra/opt/optimizers/tensor_train_optimizer.py:9]: I think they are also used here in TTOpt, so I'd think this `Bounds` class should be in a higher level of the directory tree, but am unsure where they should be.

File: tests/orquestra/opt/optimizers/tensor_train_optimizer_test.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            "n_rounds": 3,
+            "maximum_number_of_candidates": 3,
+        }
+    ]
+)
+def optimizer(request):
+    return TensorTrainOptimizer(**request.param)
+
+
+class TestTensorTrainOptimizer:
```
[the remark this answers is not in the document]
Comment (VolodyaCO) [reply tests/orquestra/opt/optimizers/tensor_train_optimizer_test.py:23]: Not for the time being. Unfortunately, the current `ttopt` library does not have any tests 😕

File: src/orquestra/opt/optimizers/tensor_train_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+                        {
+                            "cost": cost,
+                            "parameters": parameters,
+                            "bounds": _get_tighter_bounds(parameters, self.bounds),
+                        },
+                    )
+                    self.candidates.pop()
+                    self.candidates_hashes.insert(i, parameters_hash)
+                    self.candidates_hashes.pop()
+                    break
```
[the remark this answers is not in the document]
Comment (VolodyaCO) [reply src/orquestra/opt/optimizers/tensor_train_optimizer.py:139]: Definitely!

**Comment by AthenaCaesura** [comment]
Sorry I left this here for so long! Getting back on it now!

File: src/orquestra/opt/optimizers/tensor_train_optimizer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/orquestra/opt/optimizers/tensor_train_optimizer.py:9]: Perhaps the bounds type could live in the `__init__.py`? Seems like a decent place for it.

**Review by AthenaCaesura — APPROVED** [review]

**Comment by AthenaCaesura** [comment]
ok! Once tests pass you should be good to merge!

**Pull request merged**